### PR TITLE
[Feature] Add Package Execution

### PR DESCRIPTION
**Pull request by ilitteri**
## Motivation

This PR addresses the snarkVM side of https://github.com/AleoHQ/aleo/issues/356.

> Note: I left a worth-reading TODO in the code.

## Test Plan

Unit tests were added.

### CODE REVIEW

File: vm/package/execute.rs [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+
+        // Deploy the package.
+        let (execution_response, execution_transaction) = package
+            .execute::<CurrentAleo, _>(
+                None,
+                caller_private_key,
+                caller_credits,
+                Identifier::from_str("mint").unwrap(),
+                &[r0, r1],
+                &mut rand::thread_rng(),
```
Comment (ljedrz) [inline vm/package/execute.rs]: `snarkvm_utilities::TestRng` should be used here instead (it registers its seed, so it's easy to reproduce if anything goes wrong)

File: vm/package/execute.rs [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+
+        // Deploy the package.
+        let (execution_response, execution_transaction) = package
+            .execute::<CurrentAleo, _>(
+                None,
+                caller0_private_key,
+                caller0_credits,
+                Identifier::from_str("transfer").unwrap(),
+                &[r0, r1, r2],
+                &mut rand::thread_rng(),
```
Comment (ljedrz) [inline vm/package/execute.rs]: same comment as https://github.com/AleoHQ/snarkVM/pull/1140/files#r987697601

File: vm/package/execute.rs [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+            // Ensure the deployment edition matches.
+            assert_eq!(<CurrentNetwork as Network>::EDITION, execution.edition());
+            // Ensure that the number of outputs is correct.
+            assert_eq!(execution_response.outputs().len(), 1);
+            // Ensure that the execution output is correct.
+            assert!(execution_response.outputs().iter().all(|output| matches!(output, Value::Record(_record))));
+        }
+
+        // Proactively remove the temporary directory (to conserve space).
+        std::fs::remove_dir_all(directory).unwrap();
```
Comment (ljedrz) [inline vm/package/execute.rs:291]:
```suggestion
```
[`tempfile::tempdir`](https://docs.rs/tempfile/3.3.0/tempfile/fn.tempdir.html) automatically removes the temporary directory on `Drop`

File: vm/package/execute.rs [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+            // Ensure the deployment edition matches.
+            assert_eq!(<CurrentNetwork as Network>::EDITION, execution.edition());
+            // Ensure that the number of outputs is correct.
+            assert_eq!(execution_response.outputs().len(), 2);
+            // Ensure that the execution output is correct.
+            assert!(execution_response.outputs().iter().all(|output| matches!(output, Value::Record(_record))));
+        }
+
+        // Proactively remove the temporary directory (to conserve space).
+        std::fs::remove_dir_all(directory).unwrap();
```
Comment (ljedrz) [inline vm/package/execute.rs]:
```suggestion
```
same comment as https://github.com/AleoHQ/snarkVM/pull/1140/files#r987702672

File: vm/package/execute.rs (hunk 3 above)
Comment (ljedrz) [inline vm/package/execute.rs]:
```suggestion
```
[`tempfile::tempdir`](https://docs.rs/tempfile/3.3.0/tempfile/fn.tempdir.html) (used in `test_helpers::sample_package`) automatically removes the temporary directory on `Drop`

File: vm/compiler/src/ledger/transaction/serialize.rs [hunk 5]
```diff
@@ -72,15 +72,14 @@ impl<'de, N: Network> Deserialize<'de> for Transaction<N> {
...
-                        // Retrieve the additional fee, if it exists.
-                        let additional_fee = match transaction["additional_fee"].as_str() {
-                            Some(additional_fee) => {
-                                Some(serde_json::from_str(additional_fee).map_err(de::Error::custom)?)
-                            }
-                            None => None,
-                        };
+                        // TODO: This is a patch that solves the execution transaction serialization/deserialization
+                        // issue. It should be solved when https://github.com/AleoHQ/snarkVM/issues/1132 or
+                        // https://github.com/AleoHQ/snarkVM/issues/1131 are addressed.
```
Comment (howardwu) [inline vm/compiler/src/ledger/transaction/serialize.rs]:
```suggestion
```

File: vm/package/execute.rs [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+    address: Address<N>,
+    program_id: ProgramID<N>,
+}
+
+impl<N: Network> ExecutionRequest<N> {
+    /// Sends the request to the given endpoint.
+    pub fn new(transaction: Transaction<N>, address: Address<N>, program_id: ProgramID<N>) -> Result<Self> {
+        ensure!(
+            matches!(transaction, Transaction::Execute(_, _, _)),
+            "Cannot create an execution request with a deploy transaction"
```
Comment (howardwu) [inline vm/package/execute.rs]:
```suggestion
            "Cannot create an execution request with a non-execution transaction"
```
^We intend to support other transaction types in the future, hence the greater specificity.

File: vm/package/execute.rs [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+pub struct ExecuteResponse<N: Network> {
+    transaction: Transaction<N>,
+}
+
+impl<N: Network> ExecuteResponse<N> {
+    /// Initializes a new deploy response.
+    pub fn new(transaction: Transaction<N>) -> Result<Self> {
+        ensure!(
+            matches!(transaction, Transaction::Execute(_, _, _)),
+            "Cannot create an execution response with a deploy transaction"
```
Comment (howardwu) [inline vm/package/execute.rs]:
```suggestion
            "Cannot create an execution response with a non-execution transaction"
```
^We intend to support other transaction types in the future, hence the greater specificity.

File: vm/package/execute.rs (hunk 3 above)
Comment (ilitteri) [reply vm/package/execute.rs]: Thanks! I didn't know that about [tempfile::tempdir](https://docs.rs/tempfile/3.3.0/tempfile/fn.tempdir.html). I noticed that other tests use this approach, should we track this?

Comment (ilitteri) [reply vm/package/execute.rs]: Did the same for https://github.com/AleoHQ/snarkVM/pull/1139

Comment (howardwu) [reply vm/package/execute.rs]: > Thanks! I didn't know that about [tempfile::tempdir](https://docs.rs/tempfile/3.3.0/tempfile/fn.tempdir.html). I noticed that other tests use this approach, should we track this?

@ilitteri let's make sure all tests use the `tempfile::tempdir` approach, much appreciated if you could take a stab at this

Comment (ilitteri) [reply vm/package/execute.rs]: Sure thing! I'll work in that PR.

**Pull request closed without merging**